### PR TITLE
Fix the minimum energy floor

### PR DESCRIPTION
**Pull request by sbird**
The minimum energy floor was being applied inconsistently. This PR fixes that and cleans up some aspects of the cooling code.

### CODE REVIEW

File: libgadget/cooling.c [hunk 1]
```diff
@@ -67,44 +69,42 @@ double DoCooling(double redshift, double u_old, double rho, double dt, struct UV
     LambdaNet = get_heatingcooling_rate(rho, u, 1 - HYDROGEN_MASSFRAC, redshift, Z, uvbg, ne_guess);
 
     /* bracketing */
-
-    if(u - u_old - LambdaNet * dt < 0)	/* heating */
+    if(LambdaNet * dt > 0)	/* heating */
```
Comment (rainwoodman) [inline libgadget/cooling.c]: No predict any more?

File: libgadget/run.c [hunk 2]
```diff
@@ -500,7 +500,7 @@ set_units(void)
     init_cosmology(&All.CP, All.TimeIC);
 
     if(All.InitGasTemp < 0)
-        All.InitGasTemp = DMAX(All.MinGasTemp, All.CP.CMBTemperature / All.TimeInit);
+        All.InitGasTemp = All.CP.CMBTemperature / All.TimeInit;
```
Comment (rainwoodman) [inline libgadget/run.c]: One concern was at high z this would be too low and cause over-cooling problems to surface.

Comment (sbird) [reply libgadget/run.c]: Quite so! I'll revert the change.

File: libgadget/cooling.c (hunk 1 above)
Comment (sbird) [reply libgadget/cooling.c]: We have u = u_old above this so for this initial step the removed code did nothing. I can put it back though if you like.

File: libgadget/run.c (hunk 2 above)
Comment (sbird) [reply libgadget/run.c]: This is now applied in init.c

File: libgadget/cooling.c [hunk 3]
```diff
@@ -67,33 +70,35 @@ double DoCooling(double redshift, double u_old, double rho, double dt, struct UV
...
+        } while(u_lower - u_old - get_heatingcooling_rate(rho, u_lower, 1 - HYDROGEN_MASSFRAC, redshift, Z, uvbg, ne_guess) * dt > 0);
     }
 
     do
     {
         u = 0.5 * (u_lower + u_upper);
+        /* If we know that the new energy
+         * is below the minimum gas internal energy, we are done here.*/
+        if(u_upper <= MinEgySpec) {
+                u = MinEgySpec;
```
Comment (rainwoodman) [inline libgadget/cooling.c]: The indentation looks strange.

File: libgadget/density.c [hunk 4]
```diff
@@ -27,6 +27,12 @@ SPH_EntVarPred(int i)
         /*Entropy limiter for the predicted entropy: makes sure entropy stays positive. */
         if(dloga > 0 && EntVarPred < 0.5*SPHP(i).Entropy)
             EntVarPred = 0.5 * SPHP(i).Entropy;
+        /* Ensures that entropy is always positive, even if we read a bad value from disc.*/
+        if(SPHP(i).Density > 0) {
```
Comment (rainwoodman) [inline libgadget/density.c]: What about flag this and issue a warning if number of negatives is nonzero?

**Comment by sbird** [comment]
I will merge after testing

**Pull request merged**